runtime/compiler/trj9/runtime/ClassUnloadAssumption.cpp

```diff
@@ -351,6 +354,149 @@ TR_RuntimeAssumptionTable::detachFromRAT(OMR::RuntimeAssumption *assumption)
+         {
+         hashTable = _tables + kind;
+         for (size_t i = 0; i < hashTable->_spineArraySize; ++i) // for each bucket in the table
+            {
+            // Look for linked list nodes to remove until all marked nodes has been deleted
+            for (cursor = hashTable->_htSpineArray[i], prev = NULL; cursor && hashTable->_markedforDetachCount[i] > 0; cursor = next) 
+               {
+               next = cursor->getNext();
+               if (cursor->isMarkedForDetach())  // Is this assumption marked?
+                  {
+                  if (prev)
+                     prev->setNext(next);
+                  else
+                     {
+                     TR_ASSERT( hashTable->_htSpineArray[i] == cursor, "RAT spine head is not cursor!" );
+                     hashTable->_htSpineArray[i] = next;
+                     }
+                  hashTable->_markedforDetachCount[i]--;
+                  // Now release the assumption
+                  incReclaimedAssumptionCount(kind);
+                  cursor->reclaim();
+                  cursor->paint(); // RAS
+                  TR_PersistentMemory::jitPersistentFree(cursor);
+                  reclaimed++; // RAS
+                  }
+               else
+                  {
+                  prev = cursor;
+                  }
+               }
+            TR_ASSERT( hashTable->_markedforDetachCount[i]==0, "RAT detach count should be 0!" );
+            }
+         _detachPending[kind] = false;
+         }
+      }
+   assumptionTableMutex->exit();
+   TR_ASSERT( marked == reclaimed, "Did not reclaime all the marked assumptions!" );
+   marked=0;
+   }
+
+/**
+ * Mark and detach all of the assumptions in the metadata's circular linked list
+ * @param md The metadata for which to mark & detach assumptions from.
+ * @param reclaimPrePrologueAssumptions Are all assumptions free to be removed (ClassUnloading vs. recompiled)
+ * This only detaches from the metadata's linked list, not a detech from the RAT,
+ * to remove from the RAT you need to call reclaimMarkedAssumptionsFromRAT()
+ */
+void TR_RuntimeAssumptionTable::markAssumptionsAndDetach(void * md, bool reclaimPrePrologueAssumptions)
+   {
+   J9JITExceptionTable *metaData = (J9JITExceptionTable*) md;
+   OMR::RuntimeAssumption *sentry = (OMR::RuntimeAssumption*)(metaData->runtimeAssumptionList);
+   OMR::RuntimeAssumption *cursor, *next;
+
+   assumptionTableMutex->enter();
+   if (sentry)
+      {
+      int32_t numAssumptionsNotReclaimed = 0; // RAS
+      TR_ASSERT(sentry->getAssumptionKind() == RuntimeAssumptionSentinel, "First assumption must be the sentinel\n");
+      OMR::RuntimeAssumption *notReclaimedList = sentry;
+      for ( cursor=sentry->getNextAssumptionForSameJittedBody(); cursor != sentry; cursor=next)
+         {
+         next = cursor->getNextAssumptionForSameJittedBody();
+         if (!metaData || cursor->isAssumingMethod(metaData, reclaimPrePrologueAssumptions))
```

Comment:
`metaData` cannot be null here, otherwise we would have crashed when doing `metaData->runtimeAssumptionList`
In the original case metaData could have been null when a method failed compilation and the metadat was not created yet.